### PR TITLE
Add API key reference page

### LINKED ISSUE #1798: Add documentation for core APIKeys API
## This is an issue with:

- [ ] Bug (site functionality or styling)
- [ ] Errata (fix needed for doc content)
- [x] New content (guide wanted)
- [ ] Update (Add missing or refresh existing content)
- [ ] Enhancement (add new site functionality)

Add documentation describing the new core API datatype `APIKey`.
- Username
- CreatedAt (generated automatically)
- Name (generated UUID)
```
type: APIKey
api_version: core/v2
metadata:
  name: 83abef1e-e7d7-4beb-91fc-79ad90084d5b
spec:
  created_at: 1570640363
  username: admin
```

Reference:
https://docs.google.com/document/d/1wdfAhajJsXCEmC0vquifkbdP5O33HSPH9PbjFLC46b4/edit#
https://github.com/sensu/sensu-go/issues/3201
https://github.com/sensu/sensu-go/pull/3303

### PR DESCRIPTION
**Pull request by hillaryfraley**
##  Description
Adds an API key reference page to the docs 

## Motivation and Context
Closes #1798

### CODE REVIEW

**Comment by nikkictl** [comment]
Bumping this to 5.15 as well.

File: content/sensu-go/5.15/reference/apikeys.md [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+api_version: core/v2
+metadata:
+  name: my-api-key
+spec:
+  key: key-id
+  username: user1
+  groups:
+  - group1
+  - group2
+  created_at: 1569874845
```
Comment (nikkictl) [inline content/sensu-go/5.15/reference/apikeys.md]:
```
type: APIKey
api_version: core/v2
metadata:
  name: 19803eb8-36a6-4203-a225-28ec4e9f4444
spec:
  created_at: 1570732266
  username: admin
```

File: content/sensu-go/5.15/reference/apikeys.md [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+  "spec": {
+    "key": "key-id",
+    "username": "user1",
+    "groups": [
+      "group1",
+      "group2"
+      ]
+    "created_at": 1569874845
+    }
+  }
```
Comment (nikkictl) [inline content/sensu-go/5.15/reference/apikeys.md]:
```
{
  "type": "APIKey",
  "api_version": "core/v2",
  "metadata": {
    "name": "19803eb8-36a6-4203-a225-28ec4e9f4444"
  },
  "spec": {
    "created_at": 1570732266,
    "username": "admin"
  }
}
```

File: content/sensu-go/5.15/reference/apikeys.md [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+---
+
+- [API key compatibility](#api-key-compatibility)
+- [API key resource structure](#api-key-resource-structure)
+- [API key specification](#api-key-specification)
+- [Examples](#examples)
+
+API keys are long-lived authentication tokens that make it more convenient to authenticate with Sensu-adjacent applications. API keys are generated by the Sensu backend, and you can provide them to applications that want to interact with the Sensu API.
+
+Use the [APIKey API][1] to create, retrieve, update, and delete API keys.
```
Comment (nikkictl) [inline content/sensu-go/5.15/reference/apikeys.md]: APIKeys cannot be updated.

File: content/sensu-go/5.15/reference/apikeys.md [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+
+{{< /language-toggle >}}
+
+## API key specification
+
+### Top-level attributes
+
+type         | 
+-------------|------
+description  | Top-level attribute specifying the resource type. API keys should always be of type `apikeys`.
```
Comment (nikkictl) [inline content/sensu-go/5.15/reference/apikeys.md]:
```suggestion
description  | Top-level attribute specifying the resource type. API keys should always be of type `APIKey`.
```

File: content/sensu-go/5.15/reference/apikeys.md [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+required     | true
+type         | String
+example      | {{< highlight shell >}}"type": "apikeys"{{< /highlight >}}
```
Comment (nikkictl) [inline content/sensu-go/5.15/reference/apikeys.md]:
```suggestion
example      | {{< highlight shell >}}"type": "APIKey"{{< /highlight >}}
```

File: content/sensu-go/5.15/reference/apikeys.md [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+api_version  | 
+-------------|------
+description  | Top-level attribute specifying the Sensu API group and version. This attribute should always be `core/v2`.
+required     | true
+type         | String
+example      | {{< highlight shell >}}"api_version": "core/v2"{{< /highlight >}}
+
+metadata     | 
+-------------|------
+description  | Top-level collection of metadata about the API key, including the `name`. The `metadata` map is always at the top level of the check definition. This means that in `wrapped-json` and `yaml` formats, the `metadata` scope occurs outside the `spec` scope.
```
Comment (nikkictl) [inline content/sensu-go/5.15/reference/apikeys.md]: This value is a UUID randomly generated by Sensu. The user cannot provide a name for an APIKey.

File: content/sensu-go/5.15/reference/apikeys.md [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+required     | true
+type         | Map of key-value pairs
+example      | {{< highlight shell >}}"metadata": {
+  "name": "my-api-key"
```
Comment (nikkictl) [inline content/sensu-go/5.15/reference/apikeys.md]:
```suggestion
  "name": "19803eb8-36a6-4203-a225-28ec4e9f4444"
```

File: content/sensu-go/5.15/reference/apikeys.md [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+type         | Map of key-value pairs
+example      | {{< highlight shell >}}"spec": {
+    "key": "key-id",
+    "username": "user1",
+    "groups": [
+      "dev",
+      "ops"
+      ]
+    "created_at": 1569874845
+    }
```
Comment (nikkictl) [inline content/sensu-go/5.15/reference/apikeys.md]: The spec only has the created_at and username fields.
```
"spec": {
    "created_at": 1570732266,
    "username": "admin"
  }
```

File: content/sensu-go/5.15/reference/apikeys.md [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+{{< /highlight >}}
+
+### Metadata attributes
+
+| name       |      |
+-------------|------
+description  | A unique string used to identify the API key. API key names cannot contain special characters or spaces (validated with Go regex [`\A[\w\.\-]+\z`](https://regex101.com/r/zo9mQU/2)).
```
Comment (nikkictl) [inline content/sensu-go/5.15/reference/apikeys.md]: This value is a UUID generated by Sensu. The user cannot provide one.

File: content/sensu-go/5.15/reference/apikeys.md [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+required     | true
+type         | String
+example      | {{< highlight shell >}}"name": "my-api-key"{{< /highlight >}}
```
Comment (nikkictl) [inline content/sensu-go/5.15/reference/apikeys.md]:
```suggestion
example      | {{< highlight shell >}}"name": "19803eb8-36a6-4203-a225-28ec4e9f4444"{{< /highlight >}}
```

File: content/sensu-go/5.15/reference/apikeys.md [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+
+### Spec attributes
+
+| key        |      |
+-------------|------
+description  | The API key ID.
+required     | true
+type         | String
+example      | {{< highlight shell >}}"key": "key-id"{{< /highlight >}}
```
Comment (nikkictl) [inline content/sensu-go/5.15/reference/apikeys.md]: This field can be removed. It is encompassed in the `name` field.

File: content/sensu-go/5.15/reference/apikeys.md [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+
+| groups     |      |
+-------------|------
+description  | The groups the user belongs to.
+required     | false
+type         | Array
+example      | {{< highlight shell >}}"groups": [
+  "group1",
+  "group2"
+]{{< /highlight >}}
```
Comment (nikkictl) [inline content/sensu-go/5.15/reference/apikeys.md]: This field can be removed. We removed this during a later phase in the specification.

File: content/sensu-go/5.15/reference/apikeys.md [hunk 13]
```diff
@@ -0,0 +1,193 @@
...
+
+| created_at |      |
+-------------|------
+description  | The time at which the API key was created. Automatically generated when the API key is created. In UUID.
```
Comment (nikkictl) [inline content/sensu-go/5.15/reference/apikeys.md]: The created at time is not a UUID. It is a Unix timestamp.

File: content/sensu-go/5.15/reference/apikeys.md [hunk 14]
```diff
@@ -0,0 +1,193 @@
...
+api_version: core/v2
+metadata:
+  name: apikey1
+spec:
+  key: 79190ae3-ca48-4e08-80e3-97e8c89af32e
+  username: user1
+  groups:
+  - dev
+  - ops
+  created_at: 1569874845
```
Comment (nikkictl) [inline content/sensu-go/5.15/reference/apikeys.md]:
```
type: APIKey
api_version: core/v2
metadata:
  name: 19803eb8-36a6-4203-a225-28ec4e9f4444
spec:
  created_at: 1570732266
  username: admin
```

File: content/sensu-go/5.15/reference/apikeys.md [hunk 15]
```diff
@@ -0,0 +1,193 @@
...
+  "spec": {
+    "key": "79190ae3-ca48-4e08-80e3-97e8c89af32e",
+    "username": "user1",
+    "groups": [
+      "dev",
+      "ops"
+      ]
+    "created_at": 1569874845
+    }
+  }
```
Comment (nikkictl) [inline content/sensu-go/5.15/reference/apikeys.md]:
```
{
  "type": "APIKey",
  "api_version": "core/v2",
  "metadata": {
    "name": "19803eb8-36a6-4203-a225-28ec4e9f4444"
  },
  "spec": {
    "created_at": 1570732266,
    "username": "admin"
  }
}
```

**Review by nikkictl — CHANGES_REQUESTED** [review]
Happy to pair with you on this if you'd like more context! The originating issues have some example payloads as well.

File: content/sensu-go/5.15/reference/apikeys.md (hunk 3 above)
Comment (hillaryfraley) [reply content/sensu-go/5.15/reference/apikeys.md]: Removed -- thank you.

**Comment by hillaryfraley** [comment]
@nikkixdev thank you -- I incorporated all of your comments and changes and re-requested review to make sure I captured everything correctly.

File: content/sensu-go/5.15/reference/apikeys.md [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+api_version  | 
+-------------|------
+description  | Top-level attribute specifying the Sensu API group and version. This attribute should always be `core/v2`.
+required     | true
+type         | String
+example      | {{< highlight shell >}}"api_version": "core/v2"{{< /highlight >}}
+
+metadata     | 
+-------------|------
+description  | Top-level collection of metadata about the API key, including the `name`. The `metadata` map is always at the top level of the check definition. This means that in `wrapped-json` and `yaml` formats, the `metadata` scope occurs outside the `spec` scope.
```
Comment (nikkictl) [inline content/sensu-go/5.15/reference/apikeys.md]:
```suggestion
description  | Top-level collection of metadata about the API key, including the `name`. The `metadata` map is always at the top level of the API key definition. This means that in `wrapped-json` and `yaml` formats, the `metadata` scope occurs outside the `spec` scope.
```

**Review by nikkictl — COMMENTED** [review]
I think users would also benefit from reading about API keys in the API overview: https://docs.sensu.io/sensu-go/5.15/api/overview/.

**Comment by hillaryfraley** [comment]
@nikkixdev I had a separate PR for the API overview and guide (https://docs.sensu.io/sensu-go/5.15/guides/use-apikey-feature/). Cameron reviewed it late last week but if you think it needs any changes, please just let me know.

The overview: https://docs.sensu.io/sensu-go/5.15/api/overview/#authenticate-with-the-api-key-feature

The guide: https://docs.sensu.io/sensu-go/5.15/guides/use-apikey-feature/

**Review by nikkictl — APPROVED** [review]
giphy

**Pull request merged**